pkg/gitlab/client.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2024 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gitlab
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	gitlab "github.com/xanzy/go-gitlab"
+
+	"github.com/abcxyz/team-link/pkg/github"
+)
+
+// ClientProvider provides a GitLab client.
+type ClientProvider struct {
+	httpClient  *http.Client
+	instanceURL string
+	keyProvider github.KeyProvider
+}
+
+// NewGitLabClientProvider creates a new GitLabClientProvider.
+func NewGitLabClientProvider(httpClient *http.Client, instanceURL string, keyProvider github.KeyProvider) *ClientProvider {
```

Comment:
I don't like that we are using the `github` package inside the `gitlab` one. Is it possible to refactor the `KeyProvider` into a common package they can both use?

Reply:
What do you think if we move all the client/provider constructor into `pkg/client/provider.go`, in there we can have `NewGitLabClient`, `NewGoogleGroupClient`, `NewGitHubClient`

Reply:
I may be misunderstanding your suggestion, but I think Client constructors should stay in their respective packages as it is likely we will want to access package private fields/vars when constructing clients.

My suggestion was specifically about the `KeyProvider` which is a pretty simple interface that just returns a sequence of bytes and does not specifically have anything to do with `GitHub` or `GitLab`.

Reply:
Sg, thanks!

Reply:
Moved `KeyProvider` to a new package called `credentials`.